### PR TITLE
Additional endpoints

### PR DESCRIPTION
**Pull request by leviska**
Implements https://github.com/openethereum/openethereum/issues/634

### CODE REVIEW

File: bin/oe/rpc.rs [hunk 1]
```diff
@@ -35,6 +35,45 @@ pub use parity_rpc::ws::{ws, Server as WsServer};
...
+            return Err(format!("expected host:port|api0;api2, but got: {}", s));
+        }
+
+        let url: Vec<&str> = split[0].split(':').collect();
+        if url.len() != 2 {
+            return Err(format!("expected host:port|api0;api2, but got: {}", s));
+        }
+
+        let apis = if split.len() == 1 {
+            "".into()
```
Comment (rimrakhimov) [inline bin/oe/rpc.rs:64]: I believe it should be invalid not to provide at least one API to open

File: bin/oe/rpc.rs [hunk 2]
```diff
@@ -189,44 +244,84 @@ pub fn new_ws<D: rpc_apis::Dependencies>(
...
+                allowed_hosts.clone(),
+                conf.max_connections,
+                rpc::WsExtractor::new(path.clone()),
+                rpc::WsExtractor::new(path.clone()),
+                rpc::WsStats::new(deps.stats.clone()),
+                conf.max_payload,
+            ),
+        )?);
+    }
+
```
Comment (rimrakhimov) [inline bin/oe/rpc.rs:298]: Maybe it is possible to eliminate code duplication in lines 219-234;250-263 and 266-297; and 328-332;340-355 and 358-380

File: bin/oe/rpc.rs [hunk 3]
```diff
@@ -152,12 +195,24 @@ pub struct Dependencies<D: rpc_apis::Dependencies> {
     pub stats: Arc<RpcStats>,
 }
 
+fn format_ws_error(url: &str, e: Result<WsServer, rpc::ws::Error>) -> Result<WsServer, String> {
```
Comment (rimrakhimov) [inline bin/oe/rpc.rs:198]: IMO, name of the function does not reflect its meaning. I would suggest to change arguments in the following way making it exactly a formatter of ws errors
```suggestion
fn format_ws_error(url: &str, e: rpc::ws::Error) -> String 
```

File: bin/oe/rpc.rs [hunk 4]
```diff
@@ -189,44 +244,84 @@ pub fn new_ws<D: rpc_apis::Dependencies>(
...
+            handler,
+            allowed_origins.clone(),
+            allowed_hosts.clone(),
+            conf.max_connections,
+            rpc::WsExtractor::new(path.clone()),
+            rpc::WsExtractor::new(path.clone()),
+            rpc::WsStats::new(deps.stats.clone()),
+            conf.max_payload,
+        ),
+    )?);
```
Comment (rimrakhimov) [inline bin/oe/rpc.rs:263]: If `format_ws_error` arguments to be changed, then it is possible just to use `map_err` for the result of `start_ws`. Also, if code duplication could be eliminated, then we may get rid of `format_ws_error` function completely

**Comment by rimrakhimov** [comment]
@leviska Thank you for the contribution! I liked the implementation in general. 

However found some minor problems and parts which could be improved in my opinion, so added my comments and suggestions above. Besides that, it seems that the client cannot be run with networking interface set to `localhost` which is still highly desirable.

Also, I believe the original issue supposed that where would be only one additional cmd argument `--jsonrpc-additional-endpoints` which serves both `ws` and `http` endpoints. I am totally fine with the way you have divided it into two independent parts. But let ask @varasev also. What do you think about that?

**Comment by leviska** [comment]
> Besides that, it seems that the client cannot be run with networking interface set to localhost which is still highly desirable.

yeah, I found that, but it also doesn't work with `--jsonrpc-interface=localhost` (at least for me?), so I wanted to clarify: do I need to fix this only for additional endpoints, or for usual endpoint too?

**Comment by varasev** [comment]
> Besides that, it seems that the client cannot be run with networking interface set to `localhost` which is still highly desirable.

Maybe it can be run with `local` or `127.0.0.1` for the localhost case?

The docs https://openethereum.github.io/Configuring-OpenEthereum says that we need to use `local` (not `localhost`).

Overall, the `--interface` option, for example, works fine with `local`. We use it in our test setup: https://github.com/poanetwork/posdao-test-setup/blob/4ee7eb69d8ae795b0517c04afac5b39c67c953ce/config/node1.openethereum.toml#L8

> I am totally fine with the way you have divided it into two independent parts. But let ask @varasev also. What do you think about that?

I'm fine with it too.

**Comment by leviska** [comment]
> Maybe it can be run with local or 127.0.0.1 for the localhost case?

Oh yeah, I've missed that function. Now it works for additional endpoints too.

Also fixed all the review notes

File: bin/oe/rpc.rs [hunk 5]
```diff
@@ -189,76 +176,116 @@ pub fn new_ws<D: rpc_apis::Dependencies>(
...
-	}
+
+    let mut endpoints = conf.additional_endpoints;
+    endpoints.push(Endpoint {
+        interface: conf.interface,
+        port: conf.port,
+        apis: conf.apis,
+    });
+
+    let max_connections = conf.max_connections;
```
Comment (leviska) [inline bin/oe/rpc.rs:187]: need to unpack the struct here. It works without unpacking in 2021 edition, but doesn't in 2018, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a4adfd6d0ac4bf09519613391e7a1320

**Review by rimrakhimov — APPROVED** [review]
It looks great for me.
@leviska thanks for your work!

The only thing which I may propose is to rename `Endpoint` structure to `RpcEndpoint` to add clarity. Otherwise, I believe we can merge the PR

**Comment by rimrakhimov** [comment]
Also, it may be a great to add some unit tests for function parsing `Endpoint` structure

**Review by varasev — COMMENTED** [review]
@leviska I guess, for `--jsonrpc-additional-endpoints` and `--ws-additional-endpoints` the value should be in double quotes, right? Example:

["host1:port1|api1;api2,host2:port2|api3"]

At least, I observe this when using `additional_endpoints` option in config `toml` file, e.g.:

```
additional_endpoints = ["all:8641|web3,all:8642|eth"]
```

**Review by varasev — APPROVED** [review]
@rimrakhimov Can you approve this?

**Comment by rimrakhimov** [comment]
@varasev Yes, I believe we can merge it

**Review by rimrakhimov — APPROVED** [review]

**Pull request merged**